web3/encoding.nim

```diff
@@ -8,14 +8,195 @@
+  encoder.encode(bytes.len.uint64)
+  encoder.padright(bytes)
+
+proc encode(encoder: var AbiEncoder, value: string) {.raises: [AbiEncodingError]} =
+  encoder.encode(value.toBytes)
+
+proc encode(encoder: var AbiEncoder, value: distinct) {.raises: [AbiEncodingError]} =
+  type Base = distinctBase(typeof(value))
+  encoder.encode(Base(value))
+
+proc encode[T](encoder: var AbiEncoder, value: seq[T]) {.raises: [AbiEncodingError].}
+
+# When encoding a seq or an array with dynamic data, we need first
+# to encode the offsets of each element, and then write the actual data.
+template encodeCollection(encoder: var AbiEncoder, value: untyped) =
+  if isDynamic(typeof(value[0])):
+    var blocks: seq[seq[byte]] = @[]
+    # Each item here will occupy a slot of 32 bytes.
+    var offset = value.len * abiSlotSize
+
+    for element in value:
+      # Store the encoded element in order
+      # to add the data after the offsets
+      var e = AbiEncoder(output: memoryOutput())
+      e.encode(element)
+      let bytes = e.finish()
+      blocks.add(bytes)
+
+      # Encode the offset of the dynamic data
+      encoder.encode(offset.uint64)
+      offset += bytes.len
+
+    for data in blocks:
+      encoder.write(data)
+  else:
+    for element in value:
+      encoder.encode(element)
+
+proc encode[T, I](encoder: var AbiEncoder, value: array[I, T]) {.raises: [AbiEncodingError].} =
+  encodeCollection(encoder, value)
+
+proc encode[T](encoder: var AbiEncoder, value: seq[T]) {.raises: [AbiEncodingError].} =
+  # The ABI specification requires that the length of the sequence is encoded first.
+  encoder.encode(value.len.uint64)
+
+  encodeCollection(encoder, value)
+
+# When encoding a tuple, we need to handle each field separately.
+# If a field is dynamic, we encode its offset first, then the data.
+# Otherwise, we encode the field directly.
+proc encode(encoder: var AbiEncoder, tupl: tuple) {.raises: [AbiEncodingError]} =
+  var data: seq[seq[byte]] = @[]
+  # Each item here will occupy a slot of 32 bytes.
+  var offset = type(tupl).arity * abiSlotSize
+
+  for field in tupl.fields:
+    when isDynamic(typeof(field)) or (typeof(field) is tuple):
+      # Store the encoded element in order
+      # to add the data after the offsets
+      var e = AbiEncoder(output: memoryOutput())
+      e.encode(field)
+      let bytes = e.finish()
+      data.add(bytes)
+
+      # Encode the offset of the dynamic data
+      encoder.encode(offset.uint64)
+      offset += bytes.len
+    else:
+      encoder.encode(field)
+
+  for data in data:
+    encoder.write(data)
+
+proc encode*[T](_: type AbiEncoder, value: T): seq[byte] {.raises: [AbiEncodingError]} =
+  try:
+    var encoder = AbiEncoder(output: memoryOutput())
+    encoder.encode(value)
+    encoder.finish()
+  except IOError as e:
+    raise newException(AbiEncodingError, "Failed to encode value: " & e.msg)
+
+# Keep the old encode functions for compatibility
```

Comment:
These should probably be `{.deprecated.}` so we can get rid of them in a future release